# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [11]:
def addTwo(x):
  return x + 2

print(addTwo(4))

# Lambda Function syntax
# ((lambda 'one or more inputs' : 'function to execute') ('value to pass through))
a_var = 7
print((lambda x: x + 2)(a_var))

6
9


#### Saving to a Variable

In [ ]:
f_test = lambda x: x+2
f_test(4)

6

#### Multiple Inputs

In [ ]:
print((lambda x,y,z: x * y * z)(3,5,8))

x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [ ]:
def multiply(f, num):
    """f expects a lambda function
       num expects a number
    """
    return f(num)
multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [ ]:
def multiply_test(num):
    return num * 4


def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply 


f_return = returnFunc()
print(returnFunc())
print(f_return(4))

#Lambda function return from a regular function
def returnLam(b, c):
  return lambda x, a: x + a + b + c
r_lamb = returnLam(4,5)
print(r_lamb(5,5))
# ^^ 4 arguments need to be passed in, two for function and two for lambda statement

<function returnFunc.<locals>.multiply at 0x7f7295cd9c20>
8
19


#### If Statements within Lambdas

In [ ]:
# Lambda x: True if (condition) else False
f_condition = lambda num: num * 2 if num > 10 else num +2

print(f_condition(8))
print(f_condition(11))

10
22


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [13]:
f = lambda x: x ** 3
f(6)


216

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [ ]:
# map(function, iterable)
# an iterable is something you can loop through; list, dict, tuple, set, or string, etc.
# Normally,the usage of map happens with a pre-defined function, but we can use lambdas as well
def squared(num, num2):
  if num < 10 and num2 < 10:
    return num ** 2, num ** 2
  else:
    return num, num2

numbers = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]

squared_nums_map = list(map(squared, numbers, more_nums))
print(squared_nums_map)


[(16, 16), (11, 10), (20, 3), (9, 9), (15, 6)]


#### Using Lambda's with Map

In [ ]:
# map(lamba x: do something,iterable)
# using lambda in map usaully takes place on one line
squared_nums_lamb = list(map(lambda x, y: (x ** 2, y **2) if x < 10 and y < 10 else (x,y), numbers, more_nums))
print(squared_nums_lamb)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [ ]:
nums_list = [2, 5, 7, 9, 10, 4, 5, 66]
calculation_operation = list(map(lambda x: ((x * 2 ) - 1), nums_list))

print(calculation_operation)

[3, 9, 13, 17, 19, 7, 9, 131]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [ ]:
names = ['Bob', 'Andy', 'Mx', 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False

new_names_list = list(filter(a_names, names))
print(new_names_list)

['Andy', 'Angelica']


#### Using Lambda's with Filter()

In [ ]:
new_names_lambs = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lambs)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [ ]:
from statistics import mean
numbers = [2, 7, 4.2, 1.6, 9, 4.4, 4.9]

def below_mean(num):
    if num < mean(numbers):
        return True
    else:
        return False

below_mean_list = list(filter(below_mean, numbers))
print(below_mean_list) 


    

[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [ ]:
# JUST YOU USE A FOR LOOP

#### Using Lambda's with Reduce()

In [ ]:
# REMOVED FROM LECTURE

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [ ]:
# REMOVED FROM LECTURE

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [ ]:
def addNums(num):
    #Set a base case to stop recursion from going forever
    if num <= 1:
        print('addNums(1) = 1')
        return num
    else:
        print(f'addNums({num}) = {num} + addNums({num - 1})')
        return num + addNums(num - 1)  

addNums(5)  

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [ ]:
def factorial(num):
    if num <= 1:
       return 1
    else:
       return num * factorial(num -1)

factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [ ]:
def subtract_num(x):
    if x <= 1:
        print('SubNums(1) = 1')
        return x
    else:
        print(f'SubNums({x}) = {x} - SubNums({x - 1})')
        return x - subtract_num(x-1) 

subtract_num(5)




SubNums(5) = 5 - SubNums(4)
SubNums(4) = 4 - SubNums(3)
SubNums(3) = 3 - SubNums(2)
SubNums(2) = 2 - SubNums(1)
SubNums(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [ ]:
def my_range(stop, start = 0, step = 2):
   while start < stop:
     yield start
     start += step

for i in my_range(20):
  my_generator_value = i
  print(my_generator_value)



0
2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [ ]:
def number_square(start, stop = 0, step = 1):
   while stop < start:
     yield start
     start -= step

for i in number_square(10):
  my_generator_value = i**2
  print(f'{i}: {my_generator_value}')

10: 100
9: 81
8: 64
7: 49
6: 36
5: 25
4: 16
3: 9
2: 4
1: 1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

final_places = list(filter(lambda place: True if place.isspace() != True and len(place) > 0 else False, places))
print(final_places)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

def LastNameSort(names):
    names.sort(key = lambda x: x.split()[-1].lower())
    return names 
name_organizer = LastNameSort(author)


name_organizer 




['Victor aNisimov',
 'Gary A.J. Bernstein',
 'Joel Carter',
 'Andrew P. Garfield',
 'David hassELHOFF']

### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

cities, temps = map(list, zip(*places))
farenheit_conversion = list(map(lambda x, y: (x, y * 1.8 + 32), cities, temps))
print(farenheit_conversion)





[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [24]:

def fib(num):
  def fib_recursion(num):
    if num == 0:
      return 1
    elif num == 1:
      return 1
    else:
      return fib_recursion(num - 1) + fib_recursion(num - 2)
  for i in range(num +1):
    print(f'Iteration {i}: {fib_recursion(i)}')


fib(5)



# I changed code to have function within function to print the correct output instead of just numeric value.
# I thought the recursion part was correct though? I set base case with return values and then fib sequence 
# would just be f(n) = f(n-1) + f(n-2)? 

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
